Set struct size cutoff in passedByValue check to 2*sizeof_pointer

dummyunit opened the pull request:
Struct arguments are either pushed to the stack or passed in the registers. Since both methods operate on machine words it is logical to tie the maximum size of a struct argument that doesn't trigger passedByValue diagnostic to the size of the machine word.

Also guessed size of STL classes is set to 3*sizeof_pointer, this better represents reality[1] and ensures that structs containing them will still trigger passedByValue.

Main goal of this change is to make results for custom `string_view`-like classes (which contain a pointer, a `size_t` and have trivial copy constructor) not depend on `--platform` option.

[1] https://godbolt.org/g/aaV0Mh

amai2012 commented on test/testother.cpp:
```diff
@@ -1551,14 +1551,15 @@ class TestOther : public TestFixture {
               "void f(X x, Y y) {\n"
               "}");
-        ASSERT_EQUALS("[test.cpp:7]: (performance) Function parameter 'y' should be passed by reference.\n", errout.str());
+        ASSERT_EQUALS("[test.cpp:8]: (performance) Function parameter 'y' should be passed by reference.\n", errout.str());
```
Removing one test, changing implementation and add another test: it's not obvious what the change really does. IMHO I would prefer if changes to the implementation do not remove tests at all.

danmar replied:
yes it seems to me your size guesses are better than the old. if you add a new test instead of rewriting the old one I think this should go in.

dummyunit replied:
As I understand, that test was checking that size calculations correctly handle inheritance. With that in mind I have adjusted it to use new unit of size (machine word a.k.a sizeof_pointer instead of bytes) that is used in passedByValue check.

Why I changed existing test instead of adding a new one? That is because the old test would fail
on 64-bit platforms. This change effectively increased the size limit from 8 to 16 bytes on 64-bit platforms, but the old test was using absolute sizes and `2*sizeof(uint64_t)` is still not big enough by the new logic. So, keeping the old test as is isn't an option.

As for new tests, I could add tests that would check that passing `struct X { void* p1; void* p2; };` by value is ok and passing `struct Y { void* p1; void* p2; char c; };` isn't.

danmar replied:
yes inheritance is one thing it tested... I see no other test that uses uint64_t. It might be interesting to test that it will generate a warning in 32-bit platforms but not in 64-bit platforms. Same code but two different Settings::platformType should generate different results.

danmar replied:
> It might be interesting to test that it will generate a warning in 32-bit platforms but not in 64-bit platforms.

If you wonder what/how I mean.. don't hesitate to ask.

dummyunit commented:
I updated the tests, split inheritance test into multiple tests:
1. Test for inheritance which shouldn't depend on size limits: base class is too big to not trigger passedByValue and derived class by itself is too small to trigger passedByValue.
2. Tests that check that struct with the same members as in the old test triggers passedByValue on 32-bit platform, but not on 64-bit platform.

I also added new test that checks if the new size limit is used.

Hope that all old test cases are now covered.
Also let me know if commits need to rebased/squashed before integration.

danmar commented:
Thanks! I applied this with ff38cc5c138b5e0361a6ad96008e5851558ba0b9
